File: apps/windows/TeslaSync.App/feature-views/HeroGauges/HeroGauges.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews.Analytics;

/// <summary>
/// The lifecycle state a <see cref="HeroGaugesViewModel"/> can be in — the native union of the
/// loading / loaded / empty / error / stale / offline branches the surface renders. The web source
/// (web/src/features/analytics/components/analytics/HeroGauges.tsx) is a presentational child that only
/// distinguishes a skeleton (data <c>undefined</c>) from the six rendered gauges; the parent analytics
/// page owns the network-failure / freshness branches. The native feature view binds the same
/// <c>GET /analytics/fleet</c> data through a shared state holder, so it reproduces every one of those
/// branches as a visible surface — none is ever hidden. <see cref="Empty"/> mirrors a fleet with no
/// distance, no energy and no drives rather than an empty HTTP body (the endpoint always returns a
/// populated object).
/// </summary>
public enum HeroGaugesState
{
    /// <summary>Initial fetch with no cached snapshot — render the six skeleton tiles.</summary>
    Loading,

    /// <summary>A fresh snapshot (or non-stale cache) with data to show — render the six gauges.</summary>
    Loaded,

    /// <summary>The snapshot resolved but carries no distance, energy or drives — render the empty state.</summary>
    Empty,

    /// <summary>The request failed and no cached snapshot exists — render the retry affordance.</summary>
    Error,

    /// <summary>A cached snapshot older than the freshness window — render the gauges plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached snapshot remains — render the gauges plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// The categorical accent a gauge tile renders with — the native mirror of the web
/// <c>MetricCard color</c> neon hue (web/src/lib/tokens.ts <c>neonColorMap</c>). Kept WinUI-free so the
/// projection can assign and the tests can assert the per-gauge colour without a UI host; the view maps
/// each value to a themed brush at render time.
/// </summary>
public enum HeroGaugeAccent
{
    /// <summary>Cyan (web <c>cyan</c>) — the distance gauge.</summary>
    Cyan,

    /// <summary>Purple (web <c>purple</c>) — the drives gauge.</summary>
    Purple,

    /// <summary>Green (web <c>green</c>) — the energy, gas-savings and CO₂ gauges.</summary>
    Green,

    /// <summary>Amber (web <c>amber</c>) — the efficiency gauge.</summary>
    Amber,
}

/// <summary>
/// The fleet analytics rollup from <c>GET /analytics/fleet</c> consumed by the hero gauges (web
/// <c>FleetAnalytics</c> in web/src/api/types.ts). Field names mirror the Go API's snake_case JSON tags
/// (<c>total_distance_km</c>, <c>total_drives</c>, <c>total_energy_kwh</c>, <c>avg_efficiency_wh_km</c>,
/// <c>total_cost</c>); parsing is null-tolerant so a partial body never throws. Distance is kilometres
/// and efficiency is Wh/km — both converted to the user's display unit only at projection time.
/// </summary>
/// <param name="TotalDistanceKm">Fleet distance in SI-derived kilometres (web <c>total_distance_km</c>).</param>
/// <param name="TotalDrives">Total completed drives, a count (web <c>total_drives</c>).</param>
/// <param name="TotalEnergyKwh">Energy consumed in kilowatt-hours (web <c>total_energy_kwh</c>).</param>
/// <param name="AvgEfficiencyWhKm">Energy intensity in watt-hours per kilometre (web <c>avg_efficiency_wh_km</c>).</param>
/// <param name="TotalCost">Charging cost in the user's currency (web <c>total_cost</c>).</param>
public sealed record HeroFleetAnalytics(
    double TotalDistanceKm,
    double TotalDrives,
    double TotalEnergyKwh,
    double AvgEfficiencyWhKm,
    double TotalCost)
{
    /// <summary>An all-zero snapshot — the parse fallback for an absent/non-object body.</summary>
    public static HeroFleetAnalytics Empty { get; } = new(0, 0, 0, 0, 0);

    /// <summary>
    /// True when there is something worth showing — at least some distance, energy or drives. Gates the
    /// empty state (the web component always renders the six tiles; the native surface shows a friendly
    /// empty state instead of six zero tiles, per the feature-view state contract).
    /// </summary>
    public bool HasData => TotalDistanceKm > 0 || TotalEnergyKwh > 0 || TotalDrives > 0;

    /// <summary>Project a <c>GET /analytics/fleet</c> JSON object into a tolerant snapshot.</summary>
    /// <param name="element">The parsed JSON body.</param>
    /// <returns>A snapshot with every absent or non-numeric field defaulted to zero.</returns>
    public static HeroFleetAnalytics FromJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        return new HeroFleetAnalytics(
            TotalDistanceKm: GetDouble(element, "total_distance_km") ?? 0,
            TotalDrives: GetDouble(element, "total_drives") ?? 0,
            TotalEnergyKwh: GetDouble(element, "total_energy_kwh") ?? 0,
            AvgEfficiencyWhKm: GetDouble(element, "avg_efficiency_wh_km") ?? 0,
            TotalCost: GetDouble(element, "total_cost") ?? 0);
    }

    private static double? GetDouble(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v))
        {
            return null;
        }

        return v.ValueKind switch
        {
            JsonValueKind.Number when v.TryGetDouble(out var n) => n,
            JsonValueKind.String when double.TryParse(v.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var n) => n,
            _ => null,
        };
    }
}

/// <summary>
/// One projected, display-ready gauge tile consumed by the WinUI view — the native analogue of one web
/// <c>&lt;MetricCard&gt;</c>. Holds the localized label, the already-formatted value, the optional unit
/// sub-line, the resolved Fluent glyph, the categorical accent (so the tile's icon colour matches the
/// web neon hue), and a Narrator automation name. Pure data — no WinUI types.
/// </summary>
/// <param name="Label">The localized tile label (web <c>MetricCard label</c>).</param>
/// <param name="Value">The pre-formatted headline value (web <c>MetricCard value</c>).</param>
/// <param name="Subtitle">The optional unit sub-line, or <see langword="null"/> (web <c>MetricCard subtitle</c>).</param>
/// <param name="Glyph">The Segoe Fluent glyph for the tile icon (web <c>MetricCard icon</c>).</param>
/// <param name="Accent">The categorical icon accent (web <c>MetricCard color</c>).</param>
/// <param name="AutomationName">The composed Narrator name for the tile.</param>
public sealed record HeroGauge(
    string Label,
    string Value,
    string? Subtitle,
    string Glyph,
    HeroGaugeAccent Accent,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the hero gauges — the native analogue of everything the web
/// component computes before returning its grid of six <c>&lt;MetricCard&gt;</c> tiles. Holds the six
/// gauges plus the data flag. Pure data so the projection is unit-tested without a UI host.
/// </summary>
/// <param name="HasData">True when the snapshot has distance, energy or drives to show.</param>
/// <param name="Gauges">The six display-ready gauge tiles, in web order.</param>
public sealed record HeroGaugesDisplay(
    bool HasData,
    IReadOnlyList<HeroGauge> Gauges);

/// <summary>
/// Pure projection from a raw <see cref="HeroFleetAnalytics"/> to the six-gauge display model — the
/// native port of the unit conversion + gas/CO₂ heuristics in
/// web/src/features/analytics/components/analytics/HeroGauges.tsx. SI is converted to the user's display
/// unit here (and only here); every label resolves through the i18n facade. The gas-savings and CO₂
/// heuristics are tied to kilometres regardless of the display unit so the dollar/kg outputs stay stable
/// for the same fleet, exactly as the web source intends.
/// </summary>
public static class HeroGaugesProjection
{
    /// <summary>Kilometres per mile used to restate Wh/km efficiency as Wh/mi (web <c>KM_PER_MILE</c>).</summary>
    public const double KmPerMile = 1.609344;

    /// <summary>Assumed petrol consumption in litres per kilometre for the gas-savings heuristic (web <c>0.085</c>).</summary>
    public const double GasLitresPerKm = 0.085;

    /// <summary>Assumed petrol price per litre for the gas-savings heuristic (web <c>1.5</c>).</summary>
    public const double GasPricePerLitre = 1.5;

    /// <summary>Assumed avoided CO₂ in kilograms per kilometre for the CO₂ heuristic (web <c>0.12</c>).</summary>
    public const double Co2KgPerKm = 0.12;

    /// <summary>Fluent glyph for the distance gauge (web <c>MapPin</c>) — Segoe Fluent Location.</summary>
    public const string DistanceGlyph = "\uE707";

    /// <summary>Fluent glyph for the drives gauge (web <c>Car</c>).</summary>
    public const string DrivesGlyph = "\uE804";

    /// <summary>Fluent glyph for the energy gauge (web <c>Zap</c>) — Segoe Fluent LightningBolt.</summary>
    public const string EnergyGlyph = "\uE945";

    /// <summary>Fluent glyph for the efficiency gauge (web <c>Gauge</c>).</summary>
    public const string EfficiencyGlyph = "\uE950";

    /// <summary>Fluent glyph for the gas-savings gauge (web <c>DollarSign</c>) — money.</summary>
    public const string GasSavingsGlyph = "\uE1D3";

    /// <summary>Fluent glyph for the CO₂ gauge (web <c>Leaf</c>) — World/eco (no native leaf glyph).</summary>
    public const string Co2Glyph = "\uE909";

    private const string MetricEfficiencyUnit = "Wh/km";
    private const string ImperialEfficiencyUnit = "Wh/mi";
    private const string EnergyUnitLabel = "kWh";
    private const string Co2UnitLabel = "kg";

    /// <summary>Project <paramref name="data"/> for the user's <paramref name="units"/> and currency.</summary>
    /// <param name="data">The fleet analytics snapshot.</param>
    /// <param name="units">The user's unit preference (distance unit drives the conversions).</param>
    /// <param name="currencySymbol">The currency symbol for the gas-savings tile (blank falls back to "$").</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <returns>The six display-ready gauges plus the data flag.</returns>
    public static HeroGaugesDisplay Project(
        HeroFleetAnalytics data,
        UnitPref units,
        string currencySymbol,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(data);
        ArgumentNullException.ThrowIfNull(units);
        ArgumentNullException.ThrowIfNull(localizer);

        var distanceUnit = units.Distance;
        string distanceUnitLabel = UnitLabels.Label(distanceUnit);
        string efficiencyUnit = distanceUnit == DistanceUnit.Mi ? ImperialEfficiencyUnit : MetricEfficiencyUnit;
        string symbol = string.IsNullOrWhiteSpace(currencySymbol) ? "$" : currencySymbol;

        // backend total_distance_km is SI km — route through the meter-based converter so the factor
        // lives in Core.Units, not here (matches the web meter-floored helper).
        double totalDist = UnitConverters.DistanceFromSi(data.TotalDistanceKm * 1000.0, distanceUnit);
        double avgEffDisplay = distanceUnit == DistanceUnit.Mi
            ? data.AvgEfficiencyWhKm * KmPerMile
            : data.AvgEfficiencyWhKm;

        // Gas savings + CO₂ are tied to kilometres regardless of the display unit (web parity).
        double gasSavings = (data.TotalDistanceKm * GasLitresPerKm * GasPricePerLitre) - data.TotalCost;
        double co2Saved = data.TotalDistanceKm * Co2KgPerKm;

        string distanceLabel = localizer.GetString("analytics.hero.distance", "Distance");
        string drivesLabel = localizer.GetString("analytics.hero.drives", "Drives");
        string energyLabel = localizer.GetString("analytics.hero.energy", "Energy");
        string efficiencyLabel = localizer.GetString("analytics.hero.efficiency", "Efficiency");
        string gasSavingsLabel = localizer.GetString("analytics.hero.gasSavings", "Gas Savings");
        string co2Label = localizer.GetString("analytics.hero.co2Saved", "CO\u2082 Saved");

        string distanceValue = ScalarFormatters.FormatNumber(totalDist, 1);
        string drivesValue = ScalarFormatters.FormatNumber(data.TotalDrives, 0);
        string energyValue = ScalarFormatters.FormatNumber(data.TotalEnergyKwh, 1);
        string efficiencyValue = ScalarFormatters.FormatNumber(avgEffDisplay, 1);
        string gasSavingsValue = ScalarFormatters.FormatCurrency(Math.Max(gasSavings, 0), symbol, 0);
        string co2Value = ScalarFormatters.FormatNumber(co2Saved, 0);

        var gauges = new HeroGauge[]
        {
            Gauge(distanceLabel, distanceValue, distanceUnitLabel, DistanceGlyph, HeroGaugeAccent.Cyan),
            Gauge(drivesLabel, drivesValue, null, DrivesGlyph, HeroGaugeAccent.Purple),
            Gauge(energyLabel, energyValue, EnergyUnitLabel, EnergyGlyph, HeroGaugeAccent.Green),
            Gauge(efficiencyLabel, efficiencyValue, efficiencyUnit, EfficiencyGlyph, HeroGaugeAccent.Amber),
            Gauge(gasSavingsLabel, gasSavingsValue, null, GasSavingsGlyph, HeroGaugeAccent.Green),
            Gauge(co2Label, co2Value, Co2UnitLabel, Co2Glyph, HeroGaugeAccent.Green),
        };

        return new HeroGaugesDisplay(data.HasData, gauges);
    }

    private static HeroGauge Gauge(string label, string value, string? subtitle, string glyph, HeroGaugeAccent accent) =>
        new(label, value, subtitle, glyph, accent, AutomationName(label, value, subtitle));

    private static string AutomationName(string label, string value, string? subtitle) =>
        string.IsNullOrEmpty(subtitle)
            ? string.Format(CultureInfo.CurrentCulture, "{0}: {1}", label, value)
            : string.Format(CultureInfo.CurrentCulture, "{0}: {1} {2}", label, value, subtitle);
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions onto parsed
/// <c>RepositoryResult&lt;HeroFleetAnalytics&gt;</c>, preserving every freshness flag
/// (cached / refreshing / stale / offline) so the view-model can render the full state matrix. Kept
/// pure so the parse-and-preserve contract is unit-tested without a network or cache.
/// </summary>
public static class HeroGaugesResultMapper
{
    /// <summary>Parse <paramref name="raw"/>'s payload (when present) while preserving its status.</summary>
    /// <param name="raw">The raw cache-then-network emission.</param>
    /// <returns>The same emission with its JSON payload parsed into a <see cref="HeroFleetAnalytics"/>.</returns>
    public static RepositoryResult<HeroFleetAnalytics> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        HeroFleetAnalytics Parse() => raw.HasValue ? HeroFleetAnalytics.FromJson(raw.Value) : HeroFleetAnalytics.Empty;

        return raw.Status switch
        {
            LoadStatus.Loading => RepositoryResult<HeroFleetAnalytics>.Loading(),
            LoadStatus.Cached => RepositoryResult<HeroFleetAnalytics>.Cached(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Refreshing => RepositoryResult<HeroFleetAnalytics>.Refreshing(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Loaded => RepositoryResult<HeroFleetAnalytics>.Loaded(Parse(), raw.FetchedAt ?? DateTimeOffset.UtcNow),
            LoadStatus.Empty => RepositoryResult<HeroFleetAnalytics>.Empty(raw.FetchedAt),
            LoadStatus.Offline => RepositoryResult<HeroFleetAnalytics>.OfflineCached(Parse(), raw.FetchedAt!.Value, raw.Error!),
            _ => RepositoryResult<HeroFleetAnalytics>.Failure(
                raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error")),
        };
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/HeroGauges/HeroGaugesModel.kt
// Pure, framework-free model + projection for the HeroGauges feature view — the native analogue of every
// value the web component derives before returning JSX
// (web/src/features/analytics/components/analytics/HeroGauges.tsx). No Compose, no Android UI, no HTTP:
// every declaration here is exercised off-device by the :app:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// HeroGauges is a presentational surface — the web component takes `data: FleetAnalytics | undefined` as a
// prop from the owning Analytics page (which owns the TanStack `useFleetAnalytics` query) and reads three
// context hooks for display: `useTranslation` (labels), `useUnits` (distance unit), and `useFormatting`
// (currency symbol + precision). It defines exactly two render branches: `!data` -> six MetricSkeleton tiles
// (loading), and the resolved payload -> six MetricCards. The cards always render — an absent value flows
// through the web `?? 0` / `safe(...)` guards to a formatted zero, never a blank box — so "empty" is the
// all-zero resolved grid, not a hidden surface. As in the sibling SummaryStatsRow port, the cache-then-network
// states (stale / offline / fetch-error) live on the owning page, not on this presentational surface; the two
// branches above are the complete state set the web source renders, reproduced verbatim here.
//
// Unit handling floors on SI exactly as the web source does: the backend serves `total_distance_km` (SI km),
// which is bridged to metres and converted through the shared [convertDistanceFromSI] — the conversion factor
// lives in the shared units lib, never here. The gas-savings and CO2 heuristics are tied to kilometres
// regardless of the display unit (web parity), so their dollar/kg outputs stay stable for the same trip. All
// numeric rendering goes through [ChartFormat.number], the native mirror of the web `fmtNumber`/`fmtInt`.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/HeroGauges — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path. `MatchingDeclarationName` is suppressed for the co-located types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.herogauges

import io.teslasync.android.components.charts.ChartFormat
import io.teslasync.android.data.UnitPreferences
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.units.DistanceUnitPref
import io.teslasync.shared.core.units.convertDistanceFromSI
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import java.util.Locale
import kotlin.math.max

// ── Web-parity constants ──────────────────────────────────────────────────────────────────────────

/** 1 km = 1000 m exactly (web `totalDistKm * 1000` bridge to SI metres). */
private const val METERS_PER_KM = 1000.0

/** km per mile — the web `KM_PER_MILE` used to convert Wh/km into Wh/mi (Wh/mi = Wh/km * km/mi). */
private const val KM_PER_MILE = 1.609344

/** Gas-savings heuristic: assumed kWh consumed per km (web `totalDistKm * 0.085 * ...`). */
private const val GAS_SAVINGS_KWH_PER_KM = 0.085

/** Gas-savings heuristic: assumed $/kWh-equivalent gas price multiplier (web `... * 1.5`). */
private const val GAS_SAVINGS_PRICE_PER_KWH = 1.5

/** CO2 heuristic: assumed kg CO2 offset per km (web `totalDistKm * 0.12`). */
private const val CO2_KG_PER_KM = 0.12

private const val DISTANCE_DECIMALS = 1
private const val DRIVES_DECIMALS = 0
private const val ENERGY_DECIMALS = 1
private const val EFFICIENCY_DECIMALS = 1
private const val GAS_SAVINGS_DECIMALS = 0
private const val CO2_DECIMALS = 0

private const val ENERGY_UNIT = "kWh"
private const val CO2_UNIT = "kg"
private const val EFFICIENCY_UNIT_MI = "Wh/mi"
private const val EFFICIENCY_UNIT_KM = "Wh/km"

private const val DEFAULT_CURRENCY = "$"
private const val DEFAULT_PRECISION = 2
private const val DEFAULT_LOCALE_TAG = "en-US"

private const val KEY_TOTAL_DISTANCE_KM = "total_distance_km"
private const val KEY_TOTAL_COST = "total_cost"
private const val KEY_AVG_EFFICIENCY_WH_KM = "avg_efficiency_wh_km"
private const val KEY_TOTAL_DRIVES = "total_drives"
private const val KEY_TOTAL_ENERGY_KWH = "total_energy_kwh"
private const val KEY_CURRENCY_SYMBOL = "currency_symbol"
private const val KEY_DECIMAL_PRECISION = "decimal_precision"

// ── Inputs ────────────────────────────────────────────────────────────────────────────────────────

/**
 * The five `/analytics/fleet` fields the web HeroGauges reads off its `FleetAnalytics` prop, decoded from
 * the SI, snake_case wire payload. Every field defaults to `0.0` when missing or JSON-null, reproducing the
 * web optional-chaining guards (`total_distance_km ?? 0`, `safe(total_cost)`, `avg_efficiency_wh_km ?? 0`,
 * and the `fmtInt`/`fmtNumber` `safeNumber` coercion on the remaining two).
 *
 * @property totalDistanceKm SI kilometres driven in the window (web `total_distance_km`).
 * @property totalCost recorded charging cost in the window (web `total_cost`).
 * @property avgEfficiencyWhKm SI energy intensity, watt-hours per km (web `avg_efficiency_wh_km`).
 * @property totalDrives drive count in the window (web `total_drives`).
 * @property totalEnergyKwh energy used in the window, already kWh on the wire (web `total_energy_kwh`).
 */
data class FleetAnalyticsSummary(
    val totalDistanceKm: Double,
    val totalCost: Double,
    val avgEfficiencyWhKm: Double,
    val totalDrives: Double,
    val totalEnergyKwh: Double,
) {
    public companion object {
        /**
         * Decodes the raw `/analytics/fleet` [json] into a [FleetAnalyticsSummary], or `null` when the
         * payload is absent. `null` and a JSON-null collapse to `null` (the web `!data` loading branch); any
         * JSON object — including an empty `{}` or one with all-zero totals — decodes to a summary, because
         * the web `!data` guard only catches `undefined`, so a populated (even empty) object renders the
         * resolved grid. A missing or JSON-null field collapses to `0.0` (web `?? 0` / `safe`).
         */
        public fun fromJson(json: JsonElement?): FleetAnalyticsSummary? {
            val obj = json as? JsonObject ?: return null
            return FleetAnalyticsSummary(
                totalDistanceKm = obj.double(KEY_TOTAL_DISTANCE_KM),
                totalCost = obj.double(KEY_TOTAL_COST),
                avgEfficiencyWhKm = obj.double(KEY_AVG_EFFICIENCY_WH_KM),
                totalDrives = obj.double(KEY_TOTAL_DRIVES),
                totalEnergyKwh = obj.double(KEY_TOTAL_ENERGY_KWH),
            )
        }
    }
}

/**
 * The display preferences this surface resolves from the live `/settings` document — the native union of the
 * web `useUnits` read (distance unit + locale) and the web `useFormatting` reads (currency symbol +
 * precision). Resolving both from one settings document mirrors the web hooks, which both derive from
 * `useSettings`.
 *
 * @property distanceUnit the user's distance unit (web `unitPrefs.distance`); selects the Distance card unit
 *   and the Wh/mi-vs-Wh/km efficiency branch.
 * @property currencySymbol the resolved currency symbol with the web blank/whitespace -> "$" fallback applied
 *   (web `useFormatting` `currencySymbol`).
 * @property precision the default fraction digits for [HeroGaugesProjection.formatCurrency] when a call omits
 *   its own (web `useFormatting` `userPrecision`: finite, non-negative, floored, else 2).
 * @property locale the BCP-47 locale driving number grouping/separators (web `fmtNumber` global locale).
 */
data class HeroGaugesDisplayPrefs(
    val distanceUnit: DistanceUnitPref,
    val currencySymbol: String,
    val precision: Int,
    val locale: Locale,
) {
    public companion object {
        /** The metric, "$", 2-dp, en-US defaults applied before settings load (web cold-start defaults). */
        public val DEFAULT: HeroGaugesDisplayPrefs = from(null)

        /** Resolves the unit + currency + precision + locale preferences from one `/settings` document. */
        public fun from(settings: JsonElement?): HeroGaugesDisplayPrefs {
            val unitPref = UnitPreferences.fromSettings(settings)
            val obj = settings as? JsonObject
            val rawSymbol = obj.stringOrNull(KEY_CURRENCY_SYMBOL)
            return HeroGaugesDisplayPrefs(
                distanceUnit = unitPref.distance,
                currencySymbol = if (!rawSymbol.isNullOrBlank()) rawSymbol else DEFAULT_CURRENCY,
                precision = obj.precisionOrNull(KEY_DECIMAL_PRECISION) ?: DEFAULT_PRECISION,
                locale = localeFor(unitPref.locale),
            )
        }

        private fun localeFor(tag: String?): Locale = Locale.forLanguageTag(tag?.takeIf { it.isNotBlank() } ?: DEFAULT_LOCALE_TAG)
    }
}

/**
 * The six localized card labels the composable resolves once (P1/S10) and threads into the projection so the
 * render-ready [HeroGaugeCard.label]s carry no English literal. Keys map 1:1 to the web `t('analytics.hero.*')`
 * calls.
 */
data class HeroGaugesStrings(
    val distance: String,
    val drives: String,
    val energy: String,
    val efficiency: String,
    val gasSavings: String,
    val co2Saved: String,
)

/** Which authored/shared glyph a card carries — kept Compose-free so the projection is unit-tested off-device. */
enum class HeroGaugeIcon { Distance, Drives, Energy, Efficiency, GasSavings, Co2 }

/** Which design-token accent a card carries (web MetricCard `color`), resolved to a Color in the composable. */
enum class HeroGaugeAccent { Info, Power, Success, Warning }

/**
 * One fully resolved hero tile — the native analogue of a single web `<MetricCard>` invocation. Pure data
 * (no Compose types) so the whole projection is asserted off-device.
 *
 * @property label the localized card label.
 * @property value the formatted primary value (already grouped + rounded for the locale).
 * @property subtitle the unit/secondary line, or `null` when the web card omits its `subtitle`.
 * @property icon the glyph slot.
 * @property accent the design-token accent slot.
 */
data class HeroGaugeCard(
    val label: String,
    val value: String,
    val subtitle: String?,
    val icon: HeroGaugeIcon,
    val accent: HeroGaugeAccent,
)

/**
 * The fully projected, render-ready view — the native analogue of everything the web component computes
 * before returning JSX. [loading] true reproduces the web `!data` branch (the composable renders six skeleton
 * tiles and ignores [cards]); otherwise [cards] holds the six resolved tiles in web order.
 */
data class HeroGaugesDisplay(
    val loading: Boolean,
    val cards: List<HeroGaugeCard>,
)

/**
 * Pure projection from the surface's prop + display preferences to its render-ready [HeroGaugesDisplay] — a
 * 1:1 port of the derivations the web component performs. The composable resolves [HeroGaugesStrings] and
 * [HeroGaugesDisplayPrefs] from the i18n catalog and the live settings, then hands them here.
 */
object HeroGaugesProjection {
    /** The fixed number of hero tiles (web `Array.from({ length: 6 })` skeletons / six MetricCards). */
    const val CARD_COUNT: Int = 6

    /**
     * Selects the render-ready view for the given [data] (the owning page's `FleetAnalytics` prop; `null`
     * while the query is in flight), the resolved display [prefs], and the localized [strings]. Reproduces the
     * web derivations verbatim: the SI-floored distance conversion, the km-tied gas-savings (clamped at zero)
     * and CO2 heuristics, the Wh/mi-vs-Wh/km efficiency branch, and the raw kWh energy / drive count.
     */
    fun project(
        data: FleetAnalyticsSummary?,
        prefs: HeroGaugesDisplayPrefs,
        strings: HeroGaugesStrings,
    ): HeroGaugesDisplay {
        if (data == null) return HeroGaugesDisplay(loading = true, cards = emptyList())

        val locale = prefs.locale
        val isMiles = prefs.distanceUnit == DistanceUnitPref.MI
        val totalDist = convertDistanceFromSI(data.totalDistanceKm * METERS_PER_KM, prefs.distanceUnit)
        val gasSavings = data.totalDistanceKm * GAS_SAVINGS_KWH_PER_KM * GAS_SAVINGS_PRICE_PER_KWH - data.totalCost
        val co2Saved = data.totalDistanceKm * CO2_KG_PER_KM
        val avgEffDisplay = if (isMiles) data.avgEfficiencyWhKm * KM_PER_MILE else data.avgEfficiencyWhKm
        val efficiencyUnit = if (isMiles) EFFICIENCY_UNIT_MI else EFFICIENCY_UNIT_KM

        return HeroGaugesDisplay(
            loading = false,
            cards =
                listOf(
                    HeroGaugeCard(
                        label = strings.distance,
                        value = ChartFormat.number(totalDist, DISTANCE_DECIMALS, locale),
                        subtitle = prefs.distanceUnit.label,
                        icon = HeroGaugeIcon.Distance,
                        accent = HeroGaugeAccent.Info,
                    ),
                    HeroGaugeCard(
                        label = strings.drives,
                        value = ChartFormat.number(data.totalDrives, DRIVES_DECIMALS, locale),
                        subtitle = null,
                        icon = HeroGaugeIcon.Drives,
                        accent = HeroGaugeAccent.Power,
                    ),
                    HeroGaugeCard(
                        label = strings.energy,
                        value = ChartFormat.number(data.totalEnergyKwh, ENERGY_DECIMALS, locale),
                        subtitle = ENERGY_UNIT,
                        icon = HeroGaugeIcon.Energy,
                        accent = HeroGaugeAccent.Success,
                    ),
                    HeroGaugeCard(
                        label = strings.efficiency,
                        value = ChartFormat.number(avgEffDisplay, EFFICIENCY_DECIMALS, locale),
                        subtitle = efficiencyUnit,
                        icon = HeroGaugeIcon.Efficiency,
                        accent = HeroGaugeAccent.Warning,
                    ),
                    HeroGaugeCard(
                        label = strings.gasSavings,
                        value = formatCurrency(max(gasSavings, 0.0), prefs, GAS_SAVINGS_DECIMALS),
                        subtitle = null,
                        icon = HeroGaugeIcon.GasSavings,
                        accent = HeroGaugeAccent.Success,
                    ),
                    HeroGaugeCard(
                        label = strings.co2Saved,
                        value = ChartFormat.number(co2Saved, CO2_DECIMALS, locale),
                        subtitle = CO2_UNIT,
                        icon = HeroGaugeIcon.Co2,
                        accent = HeroGaugeAccent.Success,
                    ),
                ),
        )
    }

    /**
     * Formats a currency [amount] the way the web `useFormatting().formatCurrency` does — the resolved
     * [HeroGaugesDisplayPrefs.currencySymbol] followed by a grouped number via the shared [ChartFormat.number].
     * [decimals] defaults to the user's [HeroGaugesDisplayPrefs.precision] (web `decimals ?? userPrecision`);
     * the HeroGauges gas-savings card overrides it with 0, matching the web `formatCurrency(..., 0)` call.
     */
    fun formatCurrency(
        amount: Double,
        prefs: HeroGaugesDisplayPrefs,
        decimals: Int = prefs.precision,
        locale: Locale = prefs.locale,
    ): String = prefs.currencySymbol + ChartFormat.number(amount, decimals.coerceAtLeast(0), locale)
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never a distance,
 * cost, efficiency, drive count, or any other fleet figure — so a diagnostics line can never leak fleet usage.
 */
object HeroGaugesDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "HeroGauges"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}

// ── JSON decode helpers (web `?? 0` / optional-chaining parity) ─────────────────────────────────────

private fun JsonObject.double(key: String): Double = (this[key] as? JsonPrimitive)?.doubleOrNull ?: 0.0

private fun JsonObject?.stringOrNull(key: String): String? = (this?.get(key) as? JsonPrimitive)?.contentOrNull

private fun JsonObject?.precisionOrNull(key: String): Int? {
    val value = (this?.get(key) as? JsonPrimitive)?.doubleOrNull ?: return null
    return if (value.isFinite() && value >= 0) value.toInt() else null
}
